File: packages/implementations/OneLang-Regex-v0.1/native/one.rb
require 'strscan'

module One
    module Regex
        def Regex.matchFromIndex(pattern, input, offset)
            scanner = StringScanner.new input
            scanner.pos = offset
            return nil if not scanner.scan(Regexp.new pattern)
            
            result = []
            i = 0
            while true do
                curr = scanner[i]
                break if not curr
                result << curr
                i = i + 1
            end
            return result
        end
    end
end

File: packages/implementations/OneLang-Regex-v0.1/native/one.go
package one

import "strings"
import "regexp"

func Regex_MatchFromIndex(pattern string, input string, offset int) []string {
	reader := strings.NewReader(input)
	reader.Seek(int64(offset), 0)

	re := regexp.MustCompile("^" + pattern)
	matchIndices := re.FindReaderSubmatchIndex(reader)

	if matchIndices == nil {
		return nil
	}

	groupCount := len(matchIndices) / 2
	result := make([]string, groupCount)		
    for i := 0; i < groupCount; i++ {
        result[i] = input[offset + matchIndices[i * 2 + 0] : offset + matchIndices[i * 2 + 1]]
    }

    return result
}
